rust/main/submitter/src/payload_dispatcher/stages/finality_stage.rs

```diff
@@ -1 +1,533 @@
+// TODO: re-enable clippy warnings
+#![allow(dead_code)]
 
+use std::{
+    collections::{HashMap, VecDeque},
+    future::Future,
+    sync::Arc,
+    time::Duration,
+};
+
+use derive_new::new;
+use eyre::{eyre, Result};
+use futures_util::future::try_join_all;
+use tokio::{
+    sync::{mpsc, Mutex},
+    time::sleep,
+};
+use tracing::{error, info, info_span, warn, Instrument};
+
+use crate::{
+    chain_tx_adapter::DispatcherError,
+    payload::{DropReason, FullPayload, PayloadStatus},
+    payload_dispatcher::stages::utils::update_tx_status,
+    transaction::{DropReason as TxDropReason, Transaction, TransactionId, TransactionStatus},
+};
+
+use super::{
+    building_stage::BuildingStageQueue, utils::retry_until_success, PayloadDispatcherState,
+};
+
+pub type FinalityStagePool = Arc<Mutex<HashMap<TransactionId, Transaction>>>;
+
+#[derive(new)]
+struct FinalityStage {
+    pool: FinalityStagePool,
+    inclusion_stage_receiver: mpsc::Receiver<Transaction>,
+    building_stage_queue: BuildingStageQueue,
+    state: PayloadDispatcherState,
+}
+
+impl FinalityStage {
+    pub async fn run(self) {
+        let FinalityStage {
+            pool,
+            inclusion_stage_receiver,
+            building_stage_queue,
+            state,
+        } = self;
+        let futures = vec![
+            tokio::spawn(
+                Self::receive_txs(inclusion_stage_receiver, pool.clone())
+                    .instrument(info_span!("receive_txs")),
+            ),
+            tokio::spawn(
+                Self::process_txs(pool, building_stage_queue, state)
+                    .instrument(info_span!("process_txs")),
+            ),
+        ];
+        if let Err(err) = try_join_all(futures).await {
+            error!(
+                error=?err,
+                "Finality stage future panicked"
+            );
+        }
+    }
+
+    async fn receive_txs(
+        mut inclusion_stage_receiver: mpsc::Receiver<Transaction>,
+        pool: FinalityStagePool,
+    ) -> Result<(), DispatcherError> {
+        loop {
+            if let Some(tx) = inclusion_stage_receiver.recv().await {
+                pool.lock().await.insert(tx.id.clone(), tx.clone());
+                info!(?tx, "Received transaction");
+            } else {
+                sleep(Duration::from_millis(500)).await;
```

Comment:
I think that `None` return value means that the channel is closed and no more messages will be delivered. If it happens, I suppose we report a critical error.

Reply:
since we don't have the metrics set up for the MVP, I'll just log an error